src/slurm_plugin/cluster_event_publisher.py

```diff
@@ -0,0 +1,371 @@
+import logging
+from datetime import datetime, timezone
+from typing import Callable, Dict, Iterable, List
+
+from slurm_plugin.common import log_exception
+from slurm_plugin.slurm_resources import DynamicNode, SlurmNode, StaticNode
+
+logger = logging.getLogger(__name__)
+
+LAUNCH_FAILURE_GROUPING = {}
+for failure in [*SlurmNode.EC2_ICE_ERROR_CODES, "LimitedInstanceCapacity"]:
+    LAUNCH_FAILURE_GROUPING.update({failure: "ice-failures"})
+
+for failure in ["VcpuLimitExceeded"]:
+    LAUNCH_FAILURE_GROUPING.update({failure: "vcpu-limit-failures"})
+
+for failure in ["VolumeLimitExceeded", "InsufficientVolumeCapacity"]:
+    LAUNCH_FAILURE_GROUPING.update({failure: "volume-limit-failures"})
+
+for failure in ["InvalidBlockDeviceMapping"]:
+    LAUNCH_FAILURE_GROUPING.update({failure: "custom-ami-errors"})
+
+for failure in ["UnauthorizedOperation", "AccessDeniedException"]:
+    LAUNCH_FAILURE_GROUPING.update({failure: "iam-policy-errors"})
+
+
+class ClusterEventPublisher:
+    def __init__(self, event_publisher: Callable, max_list_size=100):
+        self._publish_event = event_publisher
+        self._max_list_size = max_list_size
+
+    @property
+    def publish_event(self):
+        return self._publish_event
+
+    @staticmethod
+    def timestamp():
+        return datetime.now(timezone.utc).isoformat(timespec="milliseconds")
+
+    @log_exception(logger, "publish_unhealthy_static_node_events", catch_exception=Exception, raise_on_error=False)
+    def publish_unhealthy_static_node_events(
+        self,
+        unhealthy_static_nodes: List[SlurmNode],
+        nodes_in_replacement: Iterable[str],
+        failed_nodes: Dict[str, List[str]],
+    ):
+        timestamp = ClusterEventPublisher.timestamp()
+
+        nodes_in_replacement = list(nodes_in_replacement)
+
+        self.publish_event(
```

Comment:
It seems that this event is specified for launch instances error from the run_instance/create_fleet API. What about we only publish this event only when there's failed_nodes?

Reply:
Unless the log level is less than or equal to debug level, no event should be published in the case where failed_nodes is empty, or am I missing something?